pom.xml

```diff
@@ -167,6 +167,17 @@
       <groupId>org.jenkins-ci.plugins</groupId>
       <artifactId>credentials-binding</artifactId>
     </dependency>
+    <!-- BouncyCastle API Plugin -->
+    <dependency>
+      <groupId>org.jenkins-ci.plugins</groupId>
+      <artifactId>bouncycastle-api</artifactId>
+    </dependency>
+    <!-- Apache Mina SSHD Core-->
+    <dependency>
+      <groupId>org.apache.sshd</groupId>
+      <artifactId>sshd-core</artifactId>
```

Comment:
this is likely a recipe for disaster depending on this multiple times in the jenkins ecosystem

Either a dependency on the `sshd` plugin should be taken or a library plugin extracted otherwise you're going to get weird issues when someone updates to a version with conflicting classes

Reply:
@arpoch I believe Tim has the correct idea.  I've tried using a dependency on the `sshd` plugin and it still generates class not found exception log messages when running Jenkins.  I hope to try shading the sshd-common artifact to see if that will resolve the issue and keep the git plugin somewhat independent of the sshd plugin.

Reply:
I believe this is what we are trying to do as mentioned here https://stackoverflow.com/questions/48782593/shading-a-dependency-in-maven

Reply:
Till now what I have learnt from my experiments with maven shade plugin

- Using the plugin to shade sshd-common artifact requires a separated project pom with no transitive dependency of any sshd artifact. Also the usage of modules in git plugin to achieve desired results doesn't seems to be a viable option. 
So is their another alternative to shading the artifact?

Reply:
A library plugin is an alternative

Reply:
Tim could you guide me more on what a library plugin is and how it's developed, what I have explored so far, library plugin is just like a normal plugin structurally but provides a functionality which other plugins could use by making it has there dependency. So taking our case in consideration, the library plugin developed would provide the functionality of decrypting passphrase protected private keys.

Reply:
docs are here: https://www.jenkins.io/doc/developer/plugin-development/dependencies-and-class-loading/#using-library-wrapper-plugins

library plugin normally doesn't provide code unless there's something really useful many other plugin's need.

they just bundle a dependency once so its version can be controlled and you can tell what version of a dep you will get